google/cloud/bigtable/internal/rate_limiter.h

```diff
@@ -0,0 +1,102 @@
+// Copyright 2023 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef GOOGLE_CLOUD_CPP_GOOGLE_CLOUD_BIGTABLE_INTERNAL_RATE_LIMITER_H
+#define GOOGLE_CLOUD_CPP_GOOGLE_CLOUD_BIGTABLE_INTERNAL_RATE_LIMITER_H
+
+#include "google/cloud/internal/clock.h"
+#include "google/cloud/log.h"
+#include "google/cloud/version.h"
+#include <chrono>
+#include <cstdint>
+#include <mutex>
+
+namespace google {
+namespace cloud {
+namespace bigtable_internal {
+GOOGLE_CLOUD_CPP_INLINE_NAMESPACE_BEGIN
+
+/**
+ * A threadsafe interface for rate limiting.
+ *
+ * This class does not sleep. It is the responsibility of the caller to sleep.
```

Comment:
Who is the caller? It's internally marked?

Reply:
At the moment, nobody is calling... but eventually some other internal class in the library will call this thing on behalf of `Table::BulkApply()`.

Yes, it is internal, but it is still good to document this stuff so we remember what this class does if (god forbid) we ever have to touch it again.

Reply:
Ack, agreed it is good to document it, but I wonder if the below could have a templated calling function


void rate_limit_x(std::func<bool()> condition, std::function<()> stuff) {
  RateLimiter limiter(clock, rate);
  while (condition()) {
    auto wait = limiter.acquire(1);
    std::this_thread_sleep_for(wait);
    stuff();
  }
}


So the caller doesn't need to think about calling sleep